forest/cli/src/cli/snapshot_fetch.rs

```diff
@@ -407,4 +491,64 @@ mod test {
             assert!(replace_extension_url(case.0.try_into().unwrap(), case.1).is_err())
         });
     }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn download_with_aria2_test_wrong_checksum() -> Result<()> {
+        if !is_github_action() && which::which("aria2c").is_err() {
+            return Ok(());
+        }
+
+        let (url, shutdown_tx) = serve_forest_logo()?;
+        let r = download_with_aria2(
+            &url,
+            temp_dir().as_os_str().to_str().unwrap_or_default(),
+            "sha-256=f640a228f127a7ad7c3d7c8fa4a9e95c5a2eb8d32561905d97191178ab383a64",
+        );
+        ensure!(r.is_err());
+        let err = r.unwrap_err().to_string();
+        ensure!(err.contains("Checksum validation failed"));
+        shutdown_tx.send(()).unwrap();
+        Ok(())
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn download_with_aria2_test_good_checksum() -> Result<()> {
+        if !is_github_action() && which::which("aria2c").is_err() {
+            return Ok(());
+        }
+
+        let (url, shutdown_tx) = serve_forest_logo()?;
+        download_with_aria2(
+            &url,
+            temp_dir().as_os_str().to_str().unwrap_or_default(),
+            "sha-256=124305d4602185addd53df5f39a35b2564baa3f51eb211b266af2db77844266d",
+        )?;
+        shutdown_tx.send(()).unwrap();
+        Ok(())
+    }
+
+    fn serve_forest_logo() -> Result<(String, tokio::sync::oneshot::Sender<()>)> {
+        let port = OsRng.gen_range(20000..40000);
+        let url = format!("http://127.0.0.1:{port}/forest_logo.png");
```

Comment:
I am not sure about this particular server, but in general it is not needed to randomly generate a random port, it should be handled by OS to get an available one. The chance for a clash is small, so feel free to skip this suggestion.

```
use std::net::TcpListener;

fn main() {
    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
    println!("{:?}", listener.local_addr());
}
```
`> Ok(127.0.0.1:62868)`

Reply:
That will be great, the only problem is that I have not figured out how to retrieve the random port with `axum` api to construct the logo url

Reply:
something along those lines? https://github.com/tokio-rs/axum/blob/main/axum/src/extract/connect_info.rs#L197-L207

Reply:
Fixed. Thanks for looking into it!